File: src/main/java/cz/vhromada/catalog/service/ShowService.java
package cz.vhromada.catalog.service;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import cz.vhromada.catalog.domain.Episode;
import cz.vhromada.catalog.domain.Season;
import cz.vhromada.catalog.domain.Show;
import cz.vhromada.catalog.repository.ShowRepository;
import cz.vhromada.catalog.utils.CatalogUtils;
import cz.vhromada.common.service.AbstractMovableService;
import cz.vhromada.common.utils.CollectionUtils;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.Cache;
import org.springframework.stereotype.Component;

/**
 * A class represents service for shows.
 *
 * @author Vladimir Hromada
 */
@Component("showService")
public class ShowService extends AbstractMovableService<Show> {

    /**
     * Creates a new instance of ShowService.
     *
     * @param showRepository repository for shows
     * @param cache          cache
     * @throws IllegalArgumentException if repository for shows is null
     *                                  or cache is null
     */
    @Autowired
    public ShowService(final ShowRepository showRepository, @Value("#{cacheManager.getCache('catalogCache')}") final Cache cache) {
        super(showRepository, cache, "shows");
    }

    @Override
    protected Show getCopy(final Show data) {
        final Show newShow = new Show();
        newShow.setCzechName(data.getCzechName());
        newShow.setOriginalName(data.getOriginalName());
        newShow.setCsfd(data.getCsfd());
        newShow.setImdbCode(data.getImdbCode());
        newShow.setWikiEn(data.getWikiEn());
        newShow.setWikiCz(data.getWikiCz());
        newShow.setPicture(data.getPicture());
        newShow.setNote(data.getNote());
        newShow.setPosition(data.getPosition());
        newShow.setGenres(new ArrayList<>(data.getGenres()));
        newShow.setSeasons(data.getSeasons().stream().map(CatalogUtils::duplicateSeason).collect(Collectors.toList()));

        return newShow;
    }

    @Override
    protected void updatePositions(final List<Show> data) {
        for (int i = 0; i < data.size(); i++) {
            final Show show = data.get(i);
            show.setPosition(i);
            if (show.getSeasons() == null) {
                show.setSeasons(new ArrayList<>());
            }
            final List<Season> seasons = CollectionUtils.getSortedData(show.getSeasons());
            for (int j = 0; j < seasons.size(); j++) {
                final Season season = seasons.get(j);
                season.setPosition(j);
                if (season.getEpisodes() == null) {
                    season.setEpisodes(new ArrayList<>());
                }
                final List<Episode> episodes = CollectionUtils.getSortedData(season.getEpisodes());
                for (int k = 0; k < episodes.size(); k++) {
                    final Episode episode = episodes.get(k);
                    episode.setPosition(k);
                }
            }
        }
    }

}

File: src/main/kotlin/cz/vhromada/catalog/service/ShowService.kt
package cz.vhromada.catalog.service

import cz.vhromada.catalog.domain.Season
import cz.vhromada.catalog.domain.Show
import cz.vhromada.catalog.repository.ShowRepository
import cz.vhromada.common.entity.Account
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.AbstractMovableService
import cz.vhromada.common.utils.sorted
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.Cache
import org.springframework.stereotype.Component

/**
 * A class represents service for shows.
 *
 * @author Vladimir Hromada
 */
@Component("showService")
class ShowService(
        private val showRepository: ShowRepository,
        accountProvider: AccountProvider,
        timeProvider: TimeProvider,
        @Value("#{cacheManager.getCache('catalogCache')}") cache: Cache) : AbstractMovableService<Show>(showRepository, accountProvider, timeProvider, cache, "shows") {

    override fun getAccountData(account: Account): List<Show> {
        return showRepository.findByAuditCreatedUser(account.id)
    }

    override fun getCopy(data: Show): Show {
        return data.copy(id = null, seasons = data.seasons.map { getCopy(it) }, genres = data.genres.map { it })
    }

    @Suppress("DuplicatedCode")
    override fun updatePositions(data: List<Show>) {
        val audit = getAudit()
        for (i in data.indices) {
            val show = data[i]
            show.position = i
            show.modify(audit)
            val seasons = show.seasons.sorted()
            for (j in seasons.indices) {
                val season = seasons[j]
                season.position = j
                season.modify(audit)
                val episodes = season.episodes.sorted()
                for (k in episodes.indices) {
                    val episode = episodes[k]
                    episode.position = k
                    episode.modify(audit)
                }
            }
        }
    }

    private fun getCopy(data: Season): Season {
        return data.copy(id = null, subtitles = data.subtitles.map { it }, episodes = data.episodes.map { it.copy(id = null) })
    }

}
